matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -163,296 +155,279 @@ public void setMeasurementErrorSigma(double measurementErrorSigma) {
-        int last = gpxList.size() - 1;
+    private List<ViterbiMatchEntry> createviterbiMatchEntries(List<MatchEntry> matchEntries,
+            EdgeFilter edgeFilter) {
+        final List<ViterbiMatchEntry> viterbiMatchEntries = new ArrayList<ViterbiMatchEntry>();
+        ViterbiMatchEntry lastViterbiMatchEntryAdded = null;
+        MatchEntry prevEntry = null;
+        int last = matchEntries.size() - 1;
         for (int i = 0; i <= last; i++) {
-            GPXEntry gpxEntry = gpxList.get(i);
-            if (i == 0 || i == last || distanceCalc.calcDist(
-                    prevEntry.getLat(), prevEntry.getLon(),
-                    gpxEntry.getLat(), gpxEntry.getLon()) > 2 * measurementErrorSigma) {
-                filtered.add(gpxEntry);
-                prevEntry = gpxEntry;
+            MatchEntry matchEntry = matchEntries.get(i);
+            // ignore those which are within 2 * measurementErrorSigma of the previous (though never
+            // ignore the first/last).
+            if (i == 0 || i == last
+                    || distanceCalc.calcDist(prevEntry.gpxEntry.getLat(),
+                            prevEntry.gpxEntry.getLon(), matchEntry.gpxEntry.getLat(),
+                            matchEntry.gpxEntry.getLon()) > 2 * measurementErrorSigma) {
+                lastViterbiMatchEntryAdded = new ViterbiMatchEntry(matchEntry);
+                viterbiMatchEntries.add(lastViterbiMatchEntryAdded);
+                prevEntry = matchEntry;
             } else {
-                logger.debug("Filter out GPX entry: {}", i + 1);
+                matchEntry.markAsNotUsedForMatching();
             }
         }
-        return filtered;
+        return viterbiMatchEntries;
     }
 
     /**
-     * Find the possible locations of each qpxEntry in the graph.
+     * Create candidates per map match event
      */
-    private List<List<QueryResult>> lookupGPXEntries(List<GPXEntry> gpxList,
-                                                     EdgeFilter edgeFilter) {
-
-        List<List<QueryResult>> gpxEntryLocations = new ArrayList<>();
-        for (GPXEntry gpxEntry : gpxList) {
-            gpxEntryLocations.add(locationIndex.findNClosest(gpxEntry.lat, gpxEntry.lon, edgeFilter,
-                    measurementErrorSigma));
+    private void calculateCandidatesPerEvent(List<ViterbiMatchEntry> viterbiMatchEntries,
+            List<QueryResult> allCandidateLocations, QueryGraph queryGraph) {
+
+        // first, find all of the *real* candidate locations for each event i.e. the nodes/edges
+        // that are nearby to the GPX entry location.
+        final EdgeFilter edgeFilter = new DefaultEdgeFilter(
+                algoOptions.getWeighting().getFlagEncoder());
+        final List<List<QueryResult>> candidateLocationsPerEvent = new ArrayList<List<QueryResult>>();
+        for (ViterbiMatchEntry viterbiMatchEntry : viterbiMatchEntries) {
+            // TODO: shouldn't we find those within e.g. 5 * accuracy? Otherwise we're not
```

Comment:
@karussell / @stefanholder - does that make sense? I haven't looked in details, but wouldn't the emission probability be non-zero for points outside measurementSigmaError? Not in scope of this PR, but thought it might be a bug.